include/PathUtilities.h

```diff
@@ -0,0 +1,289 @@
+*/
+/* We welcome the use of the Mezzanine engine to anyone, including companies who wish to
+   Build professional software and charge for their product.
+
+   However there are some practical restrictions, so if your project involves
+   any of the following you should contact us and we will try to work something
+   out:
+    - DRM or Copy Protection of any kind(except Copyrights)
+    - Software Patents You Do Not Wish to Freely License
+    - Any Kind of Linking to Non-GPL licensed Works
+    - Are Currently In Violation of Another Copyright Holder's GPL License
+    - If You want to change our code and not add a few hundred MB of stuff to
+        your distribution
+
+   These and other limitations could cause serious legal problems if you ignore
+   them, so it is best to simply contact us or the Free Software Foundation, if
+   you have any questions.
+
+   Joseph Toppi - toppij@gmail.com
+   John Blackwood - makoenergy02@gmail.com
+*/
+#ifndef Mezz_Filesystem_PathUtilities_h
+#define Mezz_Filesystem_PathUtilities_h
+
+#ifndef SWIG
+    #include "StringTools.h"
+#endif
+
+namespace Mezzanine {
+namespace Filesystem {
+    ///////////////////////////////////////////////////////////////////////////////
+    // Dir and Base Name
+
+    /// @brief Get the directory portion of a string.
+    /// @param PathAndFile A complete path and filename.
+    /// @return If passed "/a/b/c.txt" or "c:\windirs\crash.exe" this will return "/a/b/" or "c:\windirs\".
+    String MEZZ_LIB GetDirName(const char* PathAndFile);
+    /// @brief Get the directory portion of a string.
+    /// @param PathAndFile A complete path and filename.
+    /// @return If passed "/a/b/c.txt" or "c:\windirs\crash.exe" this will return "/a/b/" or "c:\windirs\".
+    String MEZZ_LIB GetDirName(const StringView PathAndFile);
+    /// @brief Get the filename portion of a string.
+    /// @param PathAndFile A complete path and filename.
+    /// @return If passed "/a/b/c.txt" or "c:\windirs\crash.exe" this will return "c.txt" or "crash.exe".
+    String MEZZ_LIB GetBaseName(const char* PathAndFile);
+    /// @brief Get the filename portion of a string.
+    /// @param PathAndFile A complete path and filename.
+    /// @return If passed "/a/b/c.txt" or "c:\windirs\crash.exe" this will return "c.txt" or "crash.exe".
+    String MEZZ_LIB GetBaseName(const StringView PathAndFile);
+
+    ///////////////////////////////////////////////////////////////////////////////
+    // Separators
+
+    /// @brief Checks to see if a character is used by the host OS to separate directories.
+    /// @param ToCheck The character to check.
+    /// @return Returns true if the character is '\\' on windows, or '/' on linux, false otherwise.
+    template<typename CharType, typename = std::enable_if_t< StringTools::is_char<CharType>::value > >
+    constexpr Boole IsDirectorySeparator_Host(const CharType ToCheck)
+    {
+    #ifdef MEZZ_WINDOWS
+        return ( ToCheck == '\\' );
```

Comment:
To reduce duplication shouldn't these be defined in terms of  IsDirectorySeparator_Posix and  IsDirectorySeparator_Windows?

Reply:
Actually wouldn't GetDirectorySeparator_Host be even better?

Reply:
All of these variants already exist (below the point where you commented).

Reply:
I meant that the slashes shouldn't be repeated and instead This function could be `return ToCheck == GetDirectorySeperator_Host();`.